Add CONTRIBUTING information for new issues

gmelikov opened the pull request:
I copied "Reporting a problem" section from the wiki's FAQ to issue template and added `Please review the guidelines for contributing to this repository.` info, i think it makes the issuing process easier for new people.

Example - https://github.com/gmelikov/zfs/issues/new

mmlb commented on .github/ISSUE_TEMPLATE.md:
```diff
@@ -0,0 +1,14 @@
+**What distribution you're using and the version.**
```
I would drop all the `.` they look weird when rendered.

mmlb commented on .github/ISSUE_TEMPLATE.md:
```diff
@@ -0,0 +1,14 @@
+**What distribution you're using and the version.**
+
+
+**What spl/zfs packages you're using and the version.**
```
I'd reword this and previous to be `you're` -> `are you`. You are reads like a directive, but you are asking a question.

mmlb commented on .github/CONTRIBUTING.md:
```diff
@@ -0,0 +1,25 @@
+### Reporting a problem
+
```
Paragraphs can be split at each sentence so that they are not huge one-liners.

bunder2015 commented:
"which" sounds better than "what"

gmelikov commented:
Thanks for comments, fixed.

gmelikov commented:
@behlendorf fully agree with you, i'll make a draft.

Waiting for any suggestions.

gmelikov commented:
I added the structure and some information. My comments marked as italic.

Please, write any additions here, i'll add them.

gmelikov commented:
I made some additions, you can see them [here](https://github.com/gmelikov/zfs/blob/master/.github/CONTRIBUTING.md).

I appreciate any addition about these moments:
* Pull Requests requirements
* Styleguides
* Debugging info (for example, how to profile performance issues)

I made minimal variant, which i can write by myself. I think we can merge it because of issue template, it'll probably help us with issues quality.

tuxoko commented:
@gmelikov 
Can you add in template that log file should be marked with \`\`\`
Thanks.

gmelikov commented:
@tuxoko done.

mmlb commented:
It may make sense to have put the note about the ``` in as a html comment. The reporter will see the note and not need remember to remove it. We can also add any other messages that are only pertinent to the report at initial report time.

gmelikov commented:
@mmlb your variant is well too, but i think if it will be a comment then reporter can easily skip it.
If the reporter has to delete the note, he'll 100% see it and will do as needed.

I'll appreciate someone's else opinion on this question, both variants are good for me.

mmlb commented:
Right that's possible, but I don't think it's likely. Reporter not knowing/using ``` for logs/long input is not likely to be the common scenario. It's very likely that issue reporters will either do so already or only need the gentlest reminder, otoh making the common case always remove it can get to be annoying.

I've had those feelings while working in another repo, but I think that was mostly because the prompts/questions were using h1/h2 headers instead of bold like you are, so I'm not going to push to have the ``` note to be a comment, just wanted to let it be known since I was pleasantly surprised when I saw that.

dinatale2 commented on .github/CONTRIBUTING.md:
```diff
@@ -0,0 +1,160 @@
+*First of all, thank you for taking the time to contribute!*
+
+The following is a set of guidelines for contributing to ZFS on Linux. 
+Using them you can help us to make a better work.
```
Perhaps something like `By using the following guidelines, you can help us make ZFS on Linux even better.`?

gmelikov replied:
Fixed.

dinatale2 commented on .github/CONTRIBUTING.md:
```diff
@@ -0,0 +1,160 @@
+  
+[Styleguides](#styleguides)
+ 
+  * [Coding conventions](#coding-conventions)
```
Should be `Coding Conventions`

gmelikov replied:
Fixed

dinatale2 commented on .github/CONTRIBUTING.md:
```diff
@@ -0,0 +1,160 @@
+, there is `--enable-debug` flag, which can be set to enable all ASSERTs and additional correctness tests.
+
+### Where can I ask for help?
+The best place to ask is our [mailing list](https://github.com/zfsonlinux/zfs/wiki/Mailing-Lists).
```
Should be `The mailing list is the best place to ask for help.`

gmelikov replied:
Fixed.

dinatale2 commented on .github/CONTRIBUTING.md:
```diff
@@ -0,0 +1,160 @@
+### Where can I ask for help?
+The best place to ask is our [mailing list](https://github.com/zfsonlinux/zfs/wiki/Mailing-Lists).
+
+Please *note*, our issue tracker is **only for bugs**, if you have thoughts about any 
```
Should be two sentences. `Please *note*, our issue tracker is **only for bugs**. If you have any comments or suggestions regarding any enhancements, please use our mailing list.`

gmelikov replied:
Fixed.

dinatale2 commented on .github/CONTRIBUTING.md:
```diff
@@ -0,0 +1,160 @@
+
+### Reporting Bugs
+
+You can open a new issue and search existing issues using the public
```
I would encourage people to search first. This will help cut down on duplicate issues. Maybe something like `If you run into an issue, please search our issue tracker to ensure the issue hasn't been reported before. If you are seeing a new issue, please create a new issue.` Thoughts?

dinatale2 commented on .github/CONTRIBUTING.md:
```diff
@@ -0,0 +1,160 @@
+
+Please make sure that what you're actually seeing is a bug and **not a support issue**. 
+
+If in doubt, please ask on the [mailing list](https://github.com/zfsonlinux/zfs/wiki/Mailing-Lists) 
```
Something like `Please contact use via the mailing list if you aren't certain that you are experiencing a bug.` Thoughts?

gmelikov replied:
Added it and moved to top of subsection.

dinatale2 commented on .github/CONTRIBUTING.md:
```diff
@@ -0,0 +1,160 @@
+If in doubt, please ask on the [mailing list](https://github.com/zfsonlinux/zfs/wiki/Mailing-Lists) 
+first, and if you're then asked to file an issue, do so.
+
+####When opening a new issue include this information at the top of the issue:
```
`When opening a new issue, please include the following information at the top of the issue:`?

gmelikov replied:
Thanks, fixed.

dinatale2 commented on .github/CONTRIBUTING.md:
```diff
@@ -0,0 +1,160 @@
+
+####When opening a new issue include this information at the top of the issue:
+
+* What distribution you're using and the version.
```
Maybe `What distribution (with version) you are using.`?

gmelikov replied:
Fixed.

dinatale2 commented on .github/CONTRIBUTING.md:
```diff
@@ -0,0 +1,160 @@
+####When opening a new issue include this information at the top of the issue:
+
+* What distribution you're using and the version.
+* What spl/zfs packages you're using and the version.
```
Maybe `The spl and zfs versions you are using.`?

gmelikov replied:
Added as ` The spl and zfs versions you are using, installation method (repository or manual compilation).`

dinatale2 commented on .github/CONTRIBUTING.md:
```diff
@@ -0,0 +1,160 @@
+
+* What distribution you're using and the version.
+* What spl/zfs packages you're using and the version.
+* Describe the problem you're observing.
```
`Describe the issue you are experiencing`?

dinatale2 commented on .github/CONTRIBUTING.md:
```diff
@@ -0,0 +1,160 @@
+* What distribution you're using and the version.
+* What spl/zfs packages you're using and the version.
+* Describe the problem you're observing.
+* Describe how to reproduce the problem.
```
Change `problem` to `issue`

dinatale2 commented on .github/CONTRIBUTING.md:
```diff
@@ -0,0 +1,160 @@
+* Describe how to reproduce the problem.
+* Including any warning/errors/backtraces from the system logs.
+
+When a new issue is opened it's not uncommon for a developer to request additional 
```
Need a comma and slight rewording `When a new issue is opened, it is not uncommon for developers to request additional information.`

gmelikov replied:
Fixed.

dinatale2 commented on .github/CONTRIBUTING.md:
```diff
@@ -0,0 +1,160 @@
+can resolve it. For example, providing a simple test case is always 
+exceptionally helpful.  
+
+Be prepared to work with the developer looking in to your bug in order to get 
```
`Be prepared to work with the developer investigating your issue. Your assistance is crucial in providing a quick solution.` Thoughts?

gmelikov replied:
Fixed.

dinatale2 commented on .github/CONTRIBUTING.md:
```diff
@@ -0,0 +1,160 @@
+
+### Suggesting Enhancements
+
+For now [mailing list](https://github.com/zfsonlinux/zfs/wiki/Mailing-Lists) is the best place
```
`Currently, the mailing list is the best place...`?

gmelikov replied:
Fixed.

dinatale2 commented on .github/CONTRIBUTING.md:
```diff
@@ -0,0 +1,160 @@
+For now [mailing list](https://github.com/zfsonlinux/zfs/wiki/Mailing-Lists) is the best place
+to discuss any enhancements.
+
+ZFS on Linux is stable, and the main priorities for our team are cleaning 
```
`We attempt to keep ZFS on Linux in a stable state. The current priorities are cleaning up the code, fixing any bugs, and improving performance.`? By saying `main priorities` it sounds like we aren't adding new features often. Thoughts?

gmelikov replied:
I tried to tell about current main goals. I don't want to create a situation then people are striking us by feature requests and waiting that core developers will prioritize them.

I agree, this part may be changed. I'll be happy to any thoughts.

dinatale2 commented on .github/CONTRIBUTING.md:
```diff
@@ -0,0 +1,160 @@
+
+### Pull Requests
+
+* One commit inside pull request recommended. Make 
```
`Please attempt to limit pull requests to a single commit. Be sure to squash multiple commits by performing a rebase.` Not sure you want to restrict a pull request to just a single large commit though. Thoughts?

gmelikov replied:
Of course not, that's why only recommended it and added `Several small pull requests are better than one huge. If you add pull request with several 
independent improvements, make sequential parts, small pull requests are easier to review and 
approve, it'll speed up merge process.`

Added your variant. Opened to any thoughts.

dinatale2 commented on .github/CONTRIBUTING.md:
```diff
@@ -0,0 +1,160 @@
+
+Additionally, you can use our [Test Suite](https://github.com/zfsonlinux/zfs/tree/master/tests).
+
+## Styleguides
```
`Style Guides`

dinatale2 commented on .github/CONTRIBUTING.md:
```diff
@@ -0,0 +1,160 @@
+
+### Coding conventions
+
+We use [C  Style  and  Coding  Standards  for  SunOS](http://www.cis.upenn.edu/%7Elee/06cse480/data/cstyle.ms.pdf) 
```
`We currently use the C Style and Coding Standards for SunOS as our coding convention.`

gmelikov replied:
Fixed.

kpande replied:
@gmelikov dear god, please stop flooding with these "Fixed." comments.

dinatale2 commented on .github/CONTRIBUTING.md:
```diff
@@ -0,0 +1,160 @@
+with comments and simple code.
+* If your pull request improves performance, please include some benchmarks.
+
+### Testing
```
@behlendorf, should we have a small blurb about the ZFS buildbot here?

dinatale2 commented on .github/CONTRIBUTING.md:
```diff
@@ -0,0 +1,160 @@
+
+### Debugging ZFS
+
+`zdb` utility is the main tool to debug ZFS. 
```
Need `The` out in front of this sentence. Instead of `main` maybe use `primary`?

gmelikov replied:
Thanks, fixed

dinatale2 commented on .github/CONTRIBUTING.md:
```diff
@@ -0,0 +1,160 @@
+  * [Pull Requests](#pull-requests)
+  * [Testing](#testing)
+  
+[Styleguides](#styleguides)
```
Should this be `Style Guides`?

gmelikov replied:
Fixed

dinatale2 commented on .github/CONTRIBUTING.md:
```diff
@@ -0,0 +1,160 @@
+
+**Which versions of spl/zfs packages are you using**
+
+[zfs ver, spl ver]
```
Spell out `version`

gmelikov replied:
Fixed.

dinatale2 commented on .github/CONTRIBUTING.md:
```diff
@@ -0,0 +1,160 @@
+### Testing
+
+We appreciate any help. If you are ready to use bleeding edge code, you can 
+help us to improve ZFS on Linux further by issuing any bugs and performance problems.
```
`We appreciate any help. If you are running the latest code, you can help us improve ZFS on Linux further by reporting general and performance issues you encounter.` Thoughts?

dinatale2 commented:
@gmelikov Thanks for submitting this! I think this is a very good start. I also would be happy to help you iterate on this to perfect it.

kpande commented:
this is fantastic, now we just need users to read it :)

kpande commented:
@gmelikov please stop commenting "fixed" on each request change, just push the changes. you're flooding email and IRC with unnecessary comments.

gmelikov commented:
@dinatale2 thanks for help1

@kpande sorry, saw it after finished 'fix'ing.

I'll commit changes in several minutes.

dinatale2 commented on .github/CONTRIBUTING.md:
```diff
@@ -0,0 +1,156 @@
+
+### Get ZFS
+
+You can build fresh zfs packages using 
```
`You can build zfs packages by following these instructions or install stable packages from your distribution's repository.`

dinatale2 commented on .github/CONTRIBUTING.md:
```diff
@@ -0,0 +1,156 @@
+
+The `zdb` utility is the primary tool to debug ZFS. 
+
+Additionally, see 
```
Should be `Additionally, using the `--enable-debug` flag when building zfs will enable all ASSERTs and additional correctness tests. Refer to the [build instructions](https://github.com/zfsonlinux/zfs/wiki/Building-ZFS#configure-and-build) for more information.`

dinatale2 commented:
It may also be worth mentioning somewhere in the guidelines that test cases should be provided when appropriate.

gmelikov commented on .github/CONTRIBUTING.md:
````diff
@@ -0,0 +1,156 @@
+
+Wrap log text with ```
+
+[any log, see https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#reporting-bugs]
````
I can add instructions to the issue template, but best way to do is as comments (so URL won't render). Or each user has to delete it from template each time.

Issue template section is from [guidance](https://mozillascience.github.io/working-open-workshop/contributing/), which @behlendorf proposed, and it is a good place to add more information and comments to it.

gmelikov commented:
@dinatale2 first of all, thank you!

Added your changes.

> Should move the "any log" line out and put it under the ``` wrapped text. The URL is not being rendered. Are you able to add instruction text to the issue template? If so, the issue template section may not be necessary.

I'll think what can i do with issue template to improve it.

> It may also be worth mentioning somewhere in the guidelines that test cases should be provided when appropriate

Added  it to pull request chapter.

dinatale2 commented on .github/CONTRIBUTING.md:
```diff
@@ -0,0 +1,156 @@
+approve, it'll speed up merge process.
+* The more complexity - the more time to review we need. We'll do our best, but you can help us 
+with comments and simple code.
+* If your pull request improves performance, please include some benchmarks.
```
Spoke to @behlendorf offline. Don't need to mention the PERF builder.

gmelikov commented:
@dinatale2 i added comments to ISSUE_TEMPLATE.md. If you want, i can remove issue template from CONTRIBUTING.md.

Further, i reworked README.markdown, example [here](https://github.com/gmelikov/zfs). Added contribution information with link to it and useful links. And logo.

I'm ready to make PULL_REQUEST_TEMPLATE.md too (example - [node.js](https://github.com/nodejs/node/blob/master/.github/PULL_REQUEST_TEMPLATE.md)), but i need some main points for it (tests requirements, or any other).

dinatale2 commented:
@gmelikov For the pull request template, perhaps [this](https://github.com/dinatale2/zfs/blob/pull_req_template/PULL_REQUEST_TEMPLATE.md) may be a good starting point.

gmelikov commented:
@dinatale2 thank you, the template is great, added it and rebased commits.

dinatale2 commented on .github/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -0,0 +1,30 @@
+- [ ] Bug fix (non-breaking change which fixes an issue)
+- [ ] New feature (non-breaking change which adds functionality)
+- [ ] Breaking change (fix or feature that would cause existing functionality to change)
+
```
If you feel it would be useful, I don't see why not. @behlendorf what are your thoughts?

behlendorf replied:
@gmelikov good idea, we have enough code cleanup changes where I think that makes sense.  We also see a lot of performance focused changes so an entry for that would make sense.

dinatale2 commented:
@gmelikov Go ahead and remove the issue template from the contributing guidelines. The comments in the template should be sufficient and you do mention what information we expect when reporting an issue. Having the issue template in the contributing guidelines seems redundant at this point. Thanks!

gmelikov commented:
@dinatale2 removed it. I redesigned system information in [issue template](https://github.com/gmelikov/zfs/issues/new), now it's a table (i think it'll be easier to read)

dinatale2 commented on .github/ISSUE_TEMPLATE.md:
```diff
@@ -0,0 +1,37 @@
+
+### System information
+<!--  add version after "|" character -->
+type                                | version/name
```
I don't think the headings are necessary to be honest.

gmelikov replied:
I agree, but, unfourtunately, github won't render table without header.

dinatale2 replied:
Bah. Ok, then just capitalize Type, Name, and Version.

dinatale2 commented on .github/ISSUE_TEMPLATE.md:
```diff
@@ -0,0 +1,37 @@
+Distribution version    | 
+Linux kernel                 | 
+Architecture                 | 
+ZFS                                | 
```
Make this `ZFS Version` to be clear maybe?

dinatale2 commented on .github/ISSUE_TEMPLATE.md:
```diff
@@ -0,0 +1,37 @@
+Linux kernel                 | 
+Architecture                 | 
+ZFS                                | 
+SPL                                | 
```
`SPL Version` for clarity?

dinatale2 commented on .github/CONTRIBUTING.md:
```diff
@@ -0,0 +1,135 @@
+* For large pull requests, please consider breaking down your changes into independent
+improvements or sequential parts (if possible). Small pull requests are easier to review and
+approve which speeds up the merging process.
+* The more complexity - the more time to review we need. We'll do our best, but you can help us 
```
This point might be better as `Try to keep pull requests simple. Simple code with comments is much easier to review and approve.

dinatale2 commented on .github/ISSUE_TEMPLATE.md:
```diff
@@ -0,0 +1,37 @@
+If you require more general support please use 
+our mailing list. https://github.com/zfsonlinux/zfs/wiki/Mailing-Lists
+
+Please fill in as much of the template below as you're able.
```
`Please fill in as much of the template as possible.`?

dinatale2 commented on .github/ISSUE_TEMPLATE.md:
```diff
@@ -0,0 +1,37 @@
+Thank you for reporting an issue.
+
+This issue tracker is for bugs and issues only.
+If you require more general support please use 
```
`If you need general support..`?

dinatale2 commented on .github/ISSUE_TEMPLATE.md:
```diff
@@ -0,0 +1,37 @@
+  ---                                  |     --- 
+Distribution name       | 
+Distribution version    | 
+Linux kernel                 | 
```
You may want to capitalize Name, Version, and Kernel above.

gmelikov commented:
@dinatale2 i added your changes.

gmelikov commented:
@behlendorf thank you for additions! Added your changes and appended pull request template by code cleanup and performance improvement types of changes. I moved breaking change type to bottom, it's more noticeable.

dinatale2 commented on .github/CONTRIBUTING.md:
```diff
@@ -0,0 +1,152 @@
+
+### Pull Requests
+
+* All pull requests must be based of the current master branch and apply without conflicts.
```
Should be `based on the current`

dinatale2 commented on .github/CONTRIBUTING.md:
```diff
@@ -0,0 +1,152 @@
+### Suggesting Enhancements
+
+ZFS on Linux is a widely deployed production filesystem which is under active development.
+The team's primary focus is on fixing known issues, improving performance and adding compelling new features.
```
Need a comma after performance. Should be `..known issues, improving performance, and...`

dinatale2 commented on .github/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -0,0 +1,32 @@
+<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
+- [ ] Bug fix (non-breaking change which fixes an issue)
+- [ ] New feature (non-breaking change which adds functionality)
+- [ ] Performance enhancement (non-breaking change which improve efficiency)
```
Should be `improves efficiency`

dinatale2 commented on .github/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -0,0 +1,32 @@
+- [ ] I have read the **CONTRIBUTING** document.
+- [ ] I have added tests to cover my changes.
+- [ ] All new and existing tests passed.
+- [ ] Change has been approved by a ZoL member.
```
Ditch the ZoL acronym to be consistent. Say `ZFS on Linux member`

behlendorf commented on .github/CONTRIBUTING.md:
```diff
@@ -0,0 +1,152 @@
+
+A variety of methods and tools are available to aid ZFS developers.  It's strongly recommended
+that when developing a patch the `--enable-debug` configure option should be set.
+This will enable additional correctness checks and all the ASSERTs to help quickly catch potential issues.
```
Since this markdown file will live in the repository please make sure all the lines are wrapped at 80 characters so it'll be readable as plain text.

gmelikov commented:
@dinatale2 thanks, fixed.
@behlendorf wrapped it.

I removed this text from 'ask fo help' section:
```
Please *note*, our issue tracker is **only for bugs**. If you have any comments or suggestions regarding any enhancements, please use our [mailing list](https://github.com/zfsonlinux/zfs/wiki/Mailing-Lists).
```

behlendorf commented:
Merged!  Thanks everybody!
